src/threaddeadlock.cpp

```diff
@@ -0,0 +1,718 @@
+            if (rit->first == c)
+            {
+                rit->second.ChangeWaitingToHeld();
+                break;
+            }
+        }
+    }
+}
+
+// c = the cs
+// isExclusive = is the current lock exclusive, for a recursive mutex (CCriticalSection) this value should always be
+// true
+void push_lock(void *c, const CLockLocation &locklocation, LockType type, bool isExclusive, bool fTry)
+{
+    std::lock_guard<std::mutex> lock(lockdata.dd_mutex);
+
+    LockStackEntry now = std::make_pair(c, locklocation);
+    // tid of the originating request
+    const uint64_t tid = getTid();
+    // If this is a blocking lock operation, we want to make sure that the locking order between 2 mutexes is consistent
+    // across the program
+    if (fTry)
+    {
+        // a try lock will either get it, or it wont. so just add it.
+        // if we dont get the lock this will be undone in destructor
+        AddNewLock(now, tid);
+        // AddNewWaitingLock(c, tid, isExclusive);
+        return;
+    }
+    // first check lock specific issues
+    if (type == LockType::SHARED)
+    {
+    TEST_1_SM:
+    TEST_2:
+    TEST_3:
+    TEST_4:
+        // shared mutexs cant recursively lock at all, check if we already have a lock on the mutex
+        auto it = lockdata.locksheldbythread.find(tid);
+        if (it == lockdata.locksheldbythread.end() || it->second.empty())
+        {
+            // intentionally left blank
+        }
+        else
+        {
+            for (auto &lockStackLock : it->second)
+            {
+                // if it is c we are recursively locking a non recursive mutex, there is a deadlock
+                if (lockStackLock.first == c)
+                {
+                    self_deadlock_detected(now, lockStackLock);
+                }
+            }
+        }
+    }
+    else if (type == LockType::RECRUSIVESHARED)
+    {
+    TEST_1_RSM:
+        // we cannot lock exclusive if we already hold shared, check for this scenario
+        if (isExclusive)
+        {
+            auto it = lockdata.locksheldbythread.find(tid);
+            if (it == lockdata.locksheldbythread.end() || it->second.empty())
+            {
+                // intentionally left blank
+            }
+            else
+            {
+                for (auto &lockStackLock : it->second)
+                {
+                    // if we have a lock and it isnt exclusive and we are attempting to get exclusive
+                    // then we will deadlock ourself
+                    if (lockStackLock.first == c && lockStackLock.second.GetExclusive() == false)
+                    {
+                        self_deadlock_detected(now, lockStackLock);
+                    }
+                }
+            }
+        }
+        else
+        {
+            // intentionally left blank
+        }
+    }
+    else if (type == LockType::RECURSIVE)
+    {
+        // this lock can not deadlock itself
+        // intentionally left blank
+    }
```

Comment:
Nit: `else { DbgAssert(!"unsupported lock type"); }` here maybe

Reply:
good idea, this will prevent an issue if the enum is extended. added.